lib/lint.ml

```diff
@@ -0,0 +1,310 @@
+  match err with
+  | UnnecessaryField field ->
+      Printf.sprintf
+        "Warning in %s: Unnecessary field '%s'. It is suggested to remove it."
+        pkg field
+  | UnmatchedName value ->
+      Printf.sprintf
+        "Error in %s: The field 'name' that doesn't match its context. Field \
+         'name' has value '%s' but was expected of value '%s'."
+        pkg
+        (OpamPackage.Name.to_string value)
+        (OpamPackage.Name.to_string (OpamPackage.name package))
+  | UnmatchedVersion value ->
+      Printf.sprintf
+        "Error in %s: The field 'version' that doesn't match its context. \
+         Field 'version' has value '%s' but was expected of value '%s'."
+        pkg
+        (OpamPackage.Version.to_string value)
+        (OpamPackage.Version.to_string (OpamPackage.version package))
+  | DubiousDuneSubst ->
+      Printf.sprintf
+        "Warning in %s: Dubious use of 'dune subst'. 'dune subst' should \
+         always only be called with {dev} (i.e. [\"dune\" \"subst\"] {dev}) If \
+         your opam file has been autogenerated by dune, you need to upgrade \
+         your dune-project to at least (lang dune 2.7)."
+        pkg
+  | WeakChecksum msg ->
+      Printf.sprintf
+        "Error in %s: Weak checksum algorithm(s) provided. Please use SHA-256 \
+         or SHA-512. Details: %s"
+        pkg msg
+  | PinDepends ->
+      Printf.sprintf
+        "Error in %s: pin-depends present. This is not allowed in the \
+         opam-repository."
+        pkg
+  | ExtraFiles ->
+      Printf.sprintf
+        "Error in %s: extra-files present. This is not allowed in the \
+         opam-repository. Please use extra-source instead."
+        pkg
+  | FailedToDownload msg ->
+      Printf.sprintf "Error in %s: Failed to download the archive. Details: %s"
+        pkg msg
+  | DuneProjectMissing ->
+      Printf.sprintf
+        "Warning in %s: The package seems to use dune but the dune-project \
+         file is missing."
+        pkg
+  | DuneConstraintMissing ->
+      Printf.sprintf
+        "Warning in %s: The package has a dune-project file but no explicit \
+         dependency on dune was found."
+        pkg
+  | BadDuneConstraint (dep, ver) ->
+      Printf.sprintf
+        "Error in %s: Your dune-project file indicates that this package \
+         requires at least dune %s but your opam file only requires dune >= \
+         %s. Please check which requirement is the right one, and fix the \
+         other."
+        pkg ver dep
+  | DuneIsBuild ->
+      Printf.sprintf
+        "Warning in %s: The package tagged dune as a build dependency. Due to \
+         a bug in dune (https://github.com/ocaml/dune/issues/2147) this should \
+         never be the case. Please remove the {build} tag from its filter."
+        pkg
+
+let ( // ) = Filename.concat
+
+let path_from_pkg ~repo_dir pkg =
+  repo_dir // "packages"
+  // OpamPackage.Name.to_string (OpamPackage.name pkg)
+  // OpamPackage.to_string pkg
```

Comment:
Since `path_from_pkg` is only used once, IMO it would be cleaner to make this something like

```suggestion

let opam_file_path_from_pkg ~repo_dir pkg =
  let ( // ) = Filename.concat in
  repo_dir // "packages"
  // OpamPackage.Name.to_string (OpamPackage.name pkg)
  // OpamPackage.to_string pkg
  // "opam"
```

Which will then simplify line 300 a tad.